monkestation/code/modules/blood_datum/components/advert_force_speak.dm

```diff
@@ -0,0 +1,134 @@
+/datum/component/advert_force_speak
+	dupe_mode = COMPONENT_DUPE_SOURCES
+	var/next_time = INFINITY
+	var/mob/living/speaker_implant/speaker_implant
+
+/datum/component/advert_force_speak/Initialize()
+	if(!isliving(parent))
+		return COMPONENT_INCOMPATIBLE
+
+/datum/component/advert_force_speak/on_source_add(source, delay=0)
+	. = ..()
+	next_time = min(next_time, world.time + delay)
+
+/datum/component/advert_force_speak/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_LIVING_LIFE, PROC_REF(on_life))
+
+/datum/component/advert_force_speak/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_LIVING_LIFE))
+
+/datum/component/advert_force_speak/Destroy(force)
+	. = ..()
+	if(speaker_implant != null)
+		qdel(speaker_implant)
+		speaker_implant = null
+
+/datum/component/advert_force_speak/proc/on_life(mob/living/source, seconds_per_tick, times_fired)
+	SIGNAL_HANDLER
+	if(world.time >= next_time && source.stat < UNCONSCIOUS)
+		var/delay = rand(2 MINUTES) + 4 MINUTES
+		if (HAS_TRAIT(source, TRAIT_SPONSOR_IMPLANT_SYNDI))
+			delay *= 2
+		next_time = world.time + delay
+		INVOKE_ASYNC(src, PROC_REF(speak), source)
+
+/datum/component/advert_force_speak/proc/speak(mob/living/source)
+	var/syndi = HAS_TRAIT(source, TRAIT_SPONSOR_IMPLANT_SYNDI)
+	var/implanted = syndi || HAS_TRAIT(source, TRAIT_SPONSOR_IMPLANT)
+	var/bad_speak = !source.can_speak() || HAS_TRAIT(source, TRAIT_ANXIOUS) || HAS_TRAIT(source, TRAIT_SOFTSPOKEN)
+	if (!source.can_speak_language(/datum/language/common))
+		bad_speak = TRUE
+	// else if (HAS_TRAIT(source, TRAIT_SIGN_LANG))
+	// 	bad_speak = FALSE
+
+	var/list/ad_list = syndi ? GLOB.advertisements.syndi_ads : GLOB.advertisements.nt_ads
+	var/ad_idx = rand(1, ad_list.len / 2) * 2
+	var/sponsor = ad_list[ad_idx - 1]
+	var/message = ad_list[ad_idx]
+	var/spans = syndi ? list("red", "syndi-propaganda") : list()
+
+	message = replacetext(message, "%name%", source.real_name || source.name)
+
+	if(implanted && bad_speak)
+		speaker_implant_say(source, message, sponsor)
+	else
+		source.say(message, language=/datum/language/common, forced="sponsorship ([sponsor])", spans=spans)
+
+	if(implanted)
+		var/datum/bank_account/bank_account = source.get_bank_account()
+		if(bank_account)
+			var/revenue = syndi ? -40 : 5
+			bank_account.adjust_money(revenue, "[sponsor]: Sponsorship Payment")
+			bank_account.bank_card_talk("Sponsorship Payment: [revenue] credits received from [sponsor].")
+
+/datum/component/advert_force_speak/proc/speaker_implant_say(mob/living/source, message, sponsor, spans)
+	if(speaker_implant == null)
+		speaker_implant = new()
+		speaker_implant.loc = source
```

Comment:
```suggestion
		speaker_implant = new(source)
```
might also be a good idea to just make it so `/mob/living/speaker_implant` automatically sets `owner` to `loc` during initialize

Reply:
also fwiw, use `forceMove` over directly setting `loc` in 99% of cases